File: creds.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"time"
)

// Credential handling. Read this before touching anything here:
//
// The bridge payload contains userID + the password HASH + sc, which together
// are the exact triple the game itself signs every API call with. Anyone
// holding them can act as the player. They are therefore treated as secrets
// with no exceptions:
//
//   - the listener binds loopback only (enforced in config validation)
//   - request bodies are never logged, at any level
//   - the on-disk file is 0600, and the mode is re-verified after writing
//   - String/GoString/MarshalJSON all REDACT, so a stray %v, %+v or a JSON
//     encode of any struct containing these cannot leak them. Persistence uses
//     a separate unexported type that opts in explicitly.

// Credentials is the signing triple, plus the browser session cookie.
//
// The cookie was originally accepted and discarded, on the reasoning that the
// API authenticates on userID+password+sc so a cookie would only allow fetching
// web pages. That reasoning was WRONG for endpoints under hotrods/:
// load_challenge redirects to the site's front page without one, with a correct
// signature and correct parameters. It is stored now, under exactly the same
// discipline as the rest - 0600, redacted in every stringification - because it
// is one more account secret and adds no new class of risk alongside a password
// hash.
type Credentials struct {
	UserID   string
	Password string
	SC       string
	Cookie   string
}

// Valid covers what every call needs. The cookie is NOT required here:
// get_values works without one, so a payload lacking it is still useful and
// must not be rejected.
func (c Credentials) Valid() bool {
	return c.UserID != "" && c.Password != "" && c.SC != ""
}

// String redacts. This is the safety net for accidental %s / %v / log.Print.
func (c Credentials) String() string {
	if !c.Valid() {
		return "Credentials{unset}"
	}
	return fmt.Sprintf("Credentials{UserID:%s Password:%s SC:%s Cookie:%s}",
		redact(c.UserID), redact(c.Password), redact(c.SC), redact(c.Cookie))
}

// GoString redacts, covering %#v.
func (c Credentials) GoString() string { return c.String() }

// MarshalJSON redacts, covering any JSON encode that reaches these - an API
// response, a structured log line, a debug dump. Persistence deliberately does
// NOT go through this path; see credStore.save.
func (c Credentials) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		UserID   string `json:"userID"`
		Password string `json:"password"`
		SC       string `json:"sc"`
		Cookie   string `json:"cookie"`
	}{redact(c.UserID), redact(c.Password), redact(c.SC), redact(c.Cookie)})
}

// redact keeps a short prefix so two different values are distinguishable in
// logs while staying useless to anyone reading them.
func redact(s string) string {
	if s == "" {
		return ""
	}
	if len(s) <= 4 {
		return "[redacted]"
	}
	return s[:2] + "…[redacted " + fmt.Sprint(len(s)) + "ch]"
}

// credsFile is the only representation that carries real values, and it exists
// solely so persistence has to be explicit about it.
type credsFile struct {
	SchemaVersion int       `json:"schema_version"`
	UserID        string    `json:"userID"`
	Password      string    `json:"password"`
	SC            string    `json:"sc"`
	Cookie        string    `json:"cookie,omitempty"`
	SKeyGen       string    `json:"skeygen,omitempty"`
	UpdatedAt     time.Time `json:"updated_at"`
}

const credsSchemaVersion = 1

// credStore is the process-wide holder. Safe for concurrent use: the bridge
// writes, the pollers read.
type credStore struct {
	mu        sync.RWMutex
	path      string
	creds     Credentials
	skeyGen   string // signing salt reported by the bridge; wins over config
	updatedAt time.Time
}

func newCredStore(path string) *credStore { return &credStore{path: path} }

// Load reads a previously saved file. A missing file is not an error: it just
// means we are waiting for the browser bridge. A corrupt file is quarantined
// rather than fatal, matching the state.json discipline.
func (s *credStore) Load() error {
	data, err := os.ReadFile(s.path)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	if err != nil {
		return err
	}
	var f credsFile
	if err := json.Unmarshal(data, &f); err != nil || f.SchemaVersion != credsSchemaVersion {
		quarantine := fmt.Sprintf("%s.corrupt-%d", s.path, time.Now().Unix())
		if rerr := os.Rename(s.path, quarantine); rerr == nil {
			return fmt.Errorf("credentials file unusable, moved to %s; waiting for the browser bridge", quarantine)
		}
		return fmt.Errorf("credentials file unusable and could not be moved aside: %w", err)
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.creds = Credentials{UserID: f.UserID, Password: f.Password, SC: f.SC, Cookie: f.Cookie}
	s.skeyGen, s.updatedAt = f.SKeyGen, f.UpdatedAt
	return nil
}

// Set replaces the stored credentials and persists them. Returns whether
// anything actually changed, so the caller can avoid logging a no-op on every
// bridge POST (the userscript re-posts on every page load).
func (s *credStore) Set(c Credentials, skeyGen string) (changed bool, err error) {
	if !c.Valid() {
		return false, errors.New("refusing to store an incomplete credential triple")
	}
	s.mu.Lock()
	changed = s.creds != c || (skeyGen != "" && skeyGen != s.skeyGen)
	s.creds = c
	if skeyGen != "" {
		s.skeyGen = skeyGen
	}
	s.updatedAt = time.Now()
	snapshot := credsFile{
		SchemaVersion: credsSchemaVersion,
		UserID:        c.UserID,
		Password:      c.Password,
		SC:            c.SC,
		Cookie:        c.Cookie,
		SKeyGen:       s.skeyGen,
		UpdatedAt:     s.updatedAt,
	}
	s.mu.Unlock()

	if s.path == "" {
		return changed, nil // memory-only, used by tests
	}
	return changed, s.save(snapshot)
}

// save writes atomically at 0600 and verifies the resulting mode. The verify
// matters because a pre-existing file keeps its own permissions through
// OpenFile, so a previously world-readable file would silently stay that way.
func (s *credStore) save(f credsFile) error {
	data, err := json.MarshalIndent(f, "", "  ")
	if err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Dir(s.path), 0o700); err != nil {
		return err
	}
	tmp := s.path + ".tmp"
	fh, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
	if err != nil {
		return err
	}
	if _, err := fh.Write(data); err != nil {
		fh.Close()
		os.Remove(tmp)
		return err
	}
	if err := fh.Sync(); err != nil {
		fh.Close()
		os.Remove(tmp)
		return err
	}
	if err := fh.Close(); err != nil {
		os.Remove(tmp)
		return err
	}
	if err := os.Chmod(tmp, 0o600); err != nil {
		os.Remove(tmp)
		return err
	}
	if err := os.Rename(tmp, s.path); err != nil {
		os.Remove(tmp)
		return err
	}
	st, err := os.Stat(s.path)
	if err != nil {
		return err
	}
	// Windows reports synthetic POSIX permission bits and ignores chmod. The
	// file inherits the current user's private AppData ACL instead.
	if perm := st.Mode().Perm(); runtime.GOOS != "windows" && perm != 0o600 {
		return fmt.Errorf("credentials file is mode %o, want 600", perm)
	}
	return nil
}

// Get returns the credentials plus the effective signing salt. ok is false
// until the bridge has delivered a usable triple.
func (s *credStore) Get() (c Credentials, salt string, ok bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.creds, s.skeyGen, s.creds.Valid()
}

// Salt returns the bridge-reported signing salt, empty if none. The caller
// falls back to the configured value; a reported value wins because it comes
// from live page context and therefore self-heals when the game rotates it.
func (s *credStore) Salt() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.skeyGen
}

// UpdatedAt is when the bridge last delivered a payload, for the HUD's
// "credentials stale" banner.
func (s *credStore) UpdatedAt() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.updatedAt
}

// GoString redacts, covering %#v. This is NOT redundant with String(): fmt
// cannot reach Credentials.GoString through credStore's fields because they are
// unexported (reflect's CanInterface is false for those, so fmt skips the
// Stringer/GoStringer methods and dumps raw field values). Without this,
// %#v on the store printed the password, sc and salt in the clear - caught by
// TestBridgeNeverLogsSecrets. Any future struct that holds credentials in an
// unexported field needs the same treatment.
func (s *credStore) GoString() string { return s.String() }

// String redacts, so logging the store itself is safe.
func (s *credStore) String() string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return fmt.Sprintf("credStore{path:%s creds:%s salt:%s updated:%s}",
		s.path, s.creds, redact(s.skeyGen), s.updatedAt.Format(time.RFC3339))
}

File: src/net/creds.rs
//! Credential store. The bridge payload is account-equivalent.
//!
//! Discipline:
//! - the listener binds loopback only
//! - request bodies are never logged
//! - the on-disk file is 0600 (re-verified after write)
//! - Display / Debug / JSON all redact

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fmt;
use std::fs;
use std::io::Write;
use std::path::PathBuf;
use std::sync::Mutex;

const SCHEMA_VERSION: i32 = 1;

#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Credentials {
    pub user_id: String,
    pub password: String,
    pub sc: String,
    pub cookie: String,
}

impl Credentials {
    pub fn valid(&self) -> bool {
        !self.user_id.is_empty() && !self.password.is_empty() && !self.sc.is_empty()
    }

    pub fn to_df(&self) -> crate::net::dfclient::Credentials {
        crate::net::dfclient::Credentials {
            user_id: self.user_id.clone(),
            password: self.password.clone(),
            sc: self.sc.clone(),
        }
    }
}

impl fmt::Display for Credentials {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if !self.valid() {
            return write!(f, "Credentials{{unset}}");
        }
        write!(
            f,
            "Credentials{{UserID:{} Password:{} SC:{} Cookie:{}}}",
            redact(&self.user_id),
            redact(&self.password),
            redact(&self.sc),
            redact(&self.cookie)
        )
    }
}

impl Serialize for Credentials {
    fn serialize<S: serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
        use serde::ser::SerializeStruct;
        let mut st = s.serialize_struct("Credentials", 4)?;
        st.serialize_field("userID", &redact(&self.user_id))?;
        st.serialize_field("password", &redact(&self.password))?;
        st.serialize_field("sc", &redact(&self.sc))?;
        st.serialize_field("cookie", &redact(&self.cookie))?;
        st.end()
    }
}

fn redact(s: &str) -> String {
    if s.is_empty() {
        return String::new();
    }
    if s.len() <= 4 {
        return "[redacted]".into();
    }
    format!("{}…[redacted {}ch]", &s[..2], s.len())
}

#[derive(Serialize, Deserialize)]
struct CredsFile {
    schema_version: i32,
    #[serde(rename = "userID")]
    user_id: String,
    password: String,
    sc: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    cookie: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    skeygen: String,
    updated_at: DateTime<Utc>,
}

struct Inner {
    creds: Credentials,
    salt: String,
    updated_at: Option<DateTime<Utc>>,
}

pub struct Store {
    path: PathBuf,
    inner: Mutex<Inner>,
}

impl fmt::Display for Store {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let g = self.inner.lock().unwrap();
        write!(
            f,
            "credStore{{path:{} creds:{} salt:{} updated:{}}}",
            self.path.display(),
            g.creds,
            redact(&g.salt),
            g.updated_at
                .map_or_else(|| "never".into(), |t| t.to_rfc3339())
        )
    }
}

impl fmt::Debug for Store {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{self}")
    }
}

impl Store {
    pub fn new(path: impl Into<PathBuf>) -> Self {
        Self {
            path: path.into(),
            inner: Mutex::new(Inner {
                creds: Credentials::default(),
                salt: String::new(),
                updated_at: None,
            }),
        }
    }

    pub fn load(&self) -> Result<(), String> {
        let data = match fs::read(&self.path) {
            Ok(d) => d,
            Err(err) if err.kind() == std::io::ErrorKind::NotFound => return Ok(()),
            Err(err) => return Err(err.to_string()),
        };
        let parsed: Result<CredsFile, _> = serde_json::from_slice(&data);
        let file = match parsed {
            Ok(f) if f.schema_version == SCHEMA_VERSION => f,
            _ => {
                let quarantine =
                    format!("{}.corrupt-{}", self.path.display(), Utc::now().timestamp());
                let _ = fs::rename(&self.path, &quarantine);
                return Err(format!(
                    "credentials file unusable, moved to {quarantine}; waiting for the browser bridge"
                ));
            }
        };
        let mut g = self.inner.lock().unwrap();
        g.creds = Credentials {
            user_id: file.user_id,
            password: file.password,
            sc: file.sc,
            cookie: file.cookie,
        };
        g.salt = file.skeygen;
        g.updated_at = Some(file.updated_at);
        Ok(())
    }

    /// Returns whether anything actually changed.
    pub fn set(&self, c: Credentials, skeygen: &str) -> Result<bool, String> {
        if !c.valid() {
            return Err("refusing to store an incomplete credential triple".into());
        }
        let (changed, snapshot) = {
            let mut g = self.inner.lock().unwrap();
            let mut changed = g.creds != c;
            if !skeygen.is_empty() && skeygen != g.salt {
                changed = true;
                g.salt = skeygen.to_string();
            }
            g.creds = c.clone();
            g.updated_at = Some(Utc::now());
            (
                changed,
                CredsFile {
                    schema_version: SCHEMA_VERSION,
                    user_id: c.user_id,
                    password: c.password,
                    sc: c.sc,
                    cookie: c.cookie,
                    skeygen: g.salt.clone(),
                    updated_at: g.updated_at.unwrap(),
                },
            )
        };
        if self.path.as_os_str().is_empty() {
            return Ok(changed);
        }
        self.save(&snapshot)?;
        Ok(changed)
    }

    fn save(&self, file: &CredsFile) -> Result<(), String> {
        let data = serde_json::to_vec_pretty(file).map_err(|e| e.to_string())?;
        if let Some(dir) = self.path.parent() {
            #[cfg(unix)]
            {
                use std::os::unix::fs::DirBuilderExt;
                fs::DirBuilder::new()
                    .recursive(true)
                    .mode(0o700)
                    .create(dir)
                    .map_err(|e| e.to_string())?;
            }
            #[cfg(not(unix))]
            {
                fs::create_dir_all(dir).map_err(|e| e.to_string())?;
            }
        }
        let tmp = self.path.with_extension("json.tmp");
        {
            let mut fh = fs::OpenOptions::new()
                .write(true)
                .create(true)
                .truncate(true)
                .open(&tmp)
                .map_err(|e| e.to_string())?;
            #[cfg(unix)]
            {
                use std::os::unix::fs::PermissionsExt;
                fh.set_permissions(fs::Permissions::from_mode(0o600))
                    .map_err(|e| e.to_string())?;
            }
            fh.write_all(&data).map_err(|e| e.to_string())?;
            fh.sync_all().map_err(|e| e.to_string())?;
        }
        fs::rename(&tmp, &self.path).map_err(|e| e.to_string())?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let meta = fs::metadata(&self.path).map_err(|e| e.to_string())?;
            let perm = meta.permissions().mode() & 0o777;
            if perm != 0o600 {
                return Err(format!("credentials file is mode {perm:o}, want 600"));
            }
        }
        Ok(())
    }

    pub fn get(&self) -> Option<(Credentials, String)> {
        let g = self.inner.lock().unwrap();
        if g.creds.valid() {
            Some((g.creds.clone(), g.salt.clone()))
        } else {
            None
        }
    }

    pub fn salt(&self) -> String {
        self.inner.lock().unwrap().salt.clone()
    }

    pub fn updated_at(&self) -> Option<DateTime<Utc>> {
        self.inner.lock().unwrap().updated_at
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn missing_file_is_ok() {
        let s = Store::new("/no/such/df-hud-credentials.json");
        s.load().unwrap();
        assert!(s.get().is_none());
    }

    #[test]
    fn set_persists_and_redacts() {
        let dir = tempfile();
        let path = dir.join("credentials.json");
        let s = Store::new(&path);
        let secret = "3a7bd3e2360a3d29eea436fcfb7e44c735d117c4";
        let changed = s
            .set(
                Credentials {
                    user_id: "1234567".into(),
                    password: secret.into(),
                    sc: "0f9a1c4e8b2d6f3a7c5e9b1d4f8a2c6e".into(),
                    cookie: "session=abc".into(),
                },
                "salt-from-page",
            )
            .unwrap();
        assert!(changed);
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let mode = fs::metadata(&path).unwrap().permissions().mode() & 0o777;
            assert_eq!(mode, 0o600);
        }
        let raw = fs::read_to_string(&path).unwrap();
        assert!(raw.contains("1234567"));
        assert!(raw.contains("salt-from-page"));
        let text = format!("{}", s);
        assert!(
            !text.contains(secret),
            "Display leaked the password: {text}"
        );
        assert!(text.contains("[redacted"));
        let json = serde_json::to_string(&s.get().unwrap().0).unwrap();
        assert!(!json.contains(secret), "JSON leaked the password: {json}");
    }

    #[test]
    fn corrupt_is_quarantined() {
        let dir = tempfile();
        let path = dir.join("credentials.json");
        fs::write(&path, "not json").unwrap();
        let s = Store::new(&path);
        let err = s.load().unwrap_err();
        assert!(err.contains("unusable"));
        assert!(!path.exists());
    }

    fn tempfile() -> PathBuf {
        let dir = std::env::temp_dir().join(format!(
            "df-hud-creds-{}-{}",
            std::process::id(),
            Utc::now().timestamp_nanos_opt().unwrap_or(0)
        ));
        fs::create_dir_all(&dir).unwrap();
        dir
    }
}
